File: src/main/java/chapter2/ConnectableObservableExample.java
package chapter2;

import io.reactivex.Observable;
import io.reactivex.observables.ConnectableObservable;
import java.util.concurrent.TimeUnit;


public class ConnectableObservableExample {

    public void marbleDiagram() {
        String[] dt = {"1", "3", "5"};
        Observable<String> balls = Observable.interval(100L, TimeUnit.MILLISECONDS)
                .map(Long::intValue)
                .map(i -> dt[i])
                .take(dt.length);
        ConnectableObservable<String> source = balls.publish();
        source.subscribe(data -> System.out.println("Subscriber #1 -> " + data));
        source.subscribe(data -> System.out.println("Subscriber #2 -> " + data));
        source.connect();

        sleep(250);
        source.subscribe(data -> System.out.println("Subscriber #3 -> " + data));
        sleep(50);
    }

    public static void sleep(int millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public static void main(String[] args) {
        ConnectableObservableExample demo = new ConnectableObservableExample();
        demo.marbleDiagram();
    }

}

File: src/main/kotlin/chapter2/ConnectableObservableExample_.kt
package chapter2

import io.reactivex.Observable
import java.util.concurrent.TimeUnit


class ConnectableObservableExample_ {

    fun marbleDiagram() {
        val dt = arrayOf("1", "3", "5")
        val balls = Observable.interval(100L, TimeUnit.MILLISECONDS)
                .map { it.toInt() }
                .map { dt[it] }
                .take(dt.size.toLong())
        val source = balls.publish()
        source.subscribe { println("Subscriber #1 -> $it") }
        source.subscribe { println("Subscriber #2 -> $it") }
        source.connect()

        Thread.sleep(250)
        source.subscribe { println("Subscriber #3 -> $it") }
        Thread.sleep(50)
    }

}


fun main(args: Array<String>) {
    val demo = ConnectableObservableExample_()
    demo.marbleDiagram()
}
